### DISCUSSION TITLE
Dependency Injection question

### DISCUSSION THREAD

**Discussion by Daniel-Xu** [tokio-rs/axum#2297] [Q&A] [answered]
I'm using DI similar to the example in the repo. However，I don't quite understand why we can't use `impl Trait` as handler parameter? (In the meantime, we can use impl Trait in the return type of the handler)

here is the handler
```

pub async fn get_films(State(repo): State<impl FilmRepo + Send + Sync>) -> impl IntoResponse {
    StatusCode::OK
}

```

and here is the error:

![2023-10-31 at 19 45](https://github.com/tokio-rs/axum/assets/548144/ea1c173b-0c44-4654-b4c5-b68536dc7140)

I changed the code and use dynamic dispatch, it works fine. However, there's `impl IntoResponse` in the return type, why didn't the compiler complain about it?

```
type DynFilmRepo = Arc<dyn FilmRepo + Send + Sync>;
pub async fn get_films(State(repo): State<DynFilmRepo>) -> impl IntoResponse {
    StatusCode::OK
}
```

**Comment by jplatte** [comment, marked as answer]
For the return type, the compiler can infer the concrete type based on the function body. For an `impl Trait` argument, it is simply impossible for the compiler to infer the correct type. You can still have generic handlers, but you have to use a named generic parameter, and specify it when registering the handler:

```rust
pub async fn get_films<R>(State(repo): State<R>) -> impl IntoResponse
where
    R: FilmRepo + Send + Sync
{
    StatusCode::OK
}

Router::new()
    .route("/", get_films::<FilmRepoImpl>)
```
